### PR TITLE
Fix bug in shared_allocator

### PR DESCRIPTION
**Pull request by Luthaf**
I wrote this while trying to fix the Windows tests for Chemfiles.jl. 

I found a bug in shared_alloc, which means we will want to do a 0.9.1 release with this patch.

### CODE REVIEW

**Comment by Luthaf** [comment]
Hey @frodofine! 

Do you think you will have time to review this? If not I'll merge this and finish the 0.9.1 release =). No pressure though, it can wait!

File: include/chemfiles/shared_allocator.hpp [hunk 1]
```diff
@@ -12,16 +12,19 @@
...
-
     /// Number of pointer sharing this reference. No need to use atomic
     /// references counting, as the allocator is protected by a mutex
     long count;
     /// How to delete the pointer when we are done with it
     std::function<void(void)> deleter;
 };
 
+namespace {
+    [[noreturn]] inline void UNINITIALIZED_DELETER() {
```
Comment (frodofine) [inline include/chemfiles/shared_allocator.hpp]: C++ attributes!

File: include/chemfiles/shared_allocator.hpp [hunk 2]
```diff
@@ -39,8 +42,9 @@ class shared_allocator {
     /// value of type T with the given arguments.
     template<class T, typename ... Args>
     static T* make_shared(Args&& ... args) {
+        auto instance = instance_.lock();
```
Comment (frodofine) [inline include/chemfiles/shared_allocator.hpp]: Interesting, is this what fixed the bug?

Comment (Luthaf) [reply include/chemfiles/shared_allocator.hpp]: No, this is just to make the code more precise. The main fix here is moving `pointers_.erase(it);` out of the `if (count == 0)` block in free/release. A tangential fix is not using iterators after they have been deleted ...

**Comment by Luthaf** [comment]
Thank you very much!

**Pull request merged**